Solved Optional Coding Challenge

queden opened the pull request:
Solved optional coding challenge using a array bitmap representing the day in minutes. 

_Note_: I had to merge from main repository to fix a bug in the pre-supplied code, which accounts for initial commits and extra changes in files. My changes are in [FindMeetingQuery.java](https://github.com/queden/google-step-2020/blob/bitmap_solution/walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java), [FindMeetingQueryTest.java](https://github.com/queden/google-step-2020/blob/bitmap_solution/walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java), and I created [Pair.java](https://github.com/queden/google-step-2020/blob/bitmap_solution/walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/Pair.java).

ArunChauhan commented on walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:
```diff
@@ -19,79 +18,209 @@
+    int curTime = 0;
+    int curDuration = 0;
+
+    // Tracks if there is a timeslot that is too short to work for mandatory attendees, but may be
```
Please add periods at the end of all your comments.  Also, please make sure you apply Google style to the file and reflow the comments to remove the extraneous line breaks.

queden replied:
Fixed in 52dd864

ArunChauhan commented on walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:
```diff
@@ -19,79 +18,209 @@
+
+      if (dayArr[curTime] == -1) { 
+        // If there are conflicts for mandatory attendees, skips over the minute
+        curTime++;
```
curTime++ is on all your branches so it could be made unconditional.  Better still, use a 'for' loop on 'curTime' instead of a 'while' loop, which would make it clear that you loop over each minute.

queden replied:
Thanks for pointing this out! I updated this. When I converted to the for loop and removed curTime++ from each branch, the if statement checking if `dayArr[curTime] == -1` is now empty, besides a comment. I was not sure if this is bad style (and could not find a conclusive answer). The main alternative I can think of is adding `dayArr[curTime] >= 0` as a condition to the rest of the branches and removing the empty branch.

ArunChauhan commented on walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:
```diff
@@ -19,79 +18,209 @@
+  /**
+   * Returns an array representation of a day, with each index representing a minute, and the value
+   * at each meeting representing how many optional attendees have a conflict at that time. If a
+   * minute does not work for all mandatory attendees, the value will be -1. If it is 0, it works
```
Nit: The 0 case is subsumed in non-negative values "reflect[ing] the number of optional attendees that have conflict".

queden replied:
Agreed, updated in 52dd864.

ArunChauhan commented on walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:
```diff
@@ -19,79 +18,209 @@
+      int rangeStart = eventTime.range.start();
+      int rangeEnd = eventTime.range.end();
+      for (int i = rangeStart; i < rangeEnd; i++) {
+        if (dayArr[i] != -1) {
```
One possible alternative that is more compact:

`darArr[i] += (dayArr[i] < 0) ? 0 : numToAdd;`

queden replied:
That's much better, thank you!

ArunChauhan commented on walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/Pair.java:
```diff
@@ -0,0 +1,57 @@
+      new Comparator<Pair>() {
+        @Override
+        public int compare(Pair a, Pair b) {
+          return -1 * Integer.compare(a.getAttendees(), b.getAttendees());
```
Is this equivalent to Integer.compare(b.getAttendees(), a.getAttendees())?  If so, that would avoid having to multiply by -1.

queden replied:
Yes, that is the same. Updated in 52dd864, thanks!

queden commented:
There were not, so I just added those tests. Thank you for pointing this out, as it originally wasn't allowing a 30 minute meeting at the end of the day. Found and updated what was causing this.

queden commented on walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:
```diff
@@ -19,79 +18,209 @@
+    // Finds the events that allow for the most optional attendees to attend
+    if (possibleTimes.size() == 1) {
+      optimalTimes.add(possibleTimes.get(0).getTimeRange());
+    } else if (possibleTimes.size() > 0) {
```
Yeah, that would be better. Just fixed in a67e371.
